src/microbe_stage/organelle_unlocks/UnlockProgress.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using UnlockConstraints;
+
+/// <summary>
+///   Stores which organelles have been unlocked by the player.
+/// </summary>
+[UseThriveSerializer]
+public class UnlockProgress
+{
+    /// <summary>
+    ///   The organelles that the player can use.
+    ///   The default value is the by default unlocked organelles (ones without a condition)
+    /// </summary>
+    [JsonProperty]
+    private readonly HashSet<OrganelleDefinition> unlockedOrganelles =
+        SimulationParameters.Instance.GetAllOrganelles().Where(organelle => organelle.UnlockConditions == null)
+            .ToHashSet();
+
+    /// <summary>
+    ///   Organelles unlocked singe the last time in the editor.
+    /// </summary>
+    [JsonProperty]
+    private readonly HashSet<OrganelleDefinition> recentlyUnlocked = new();
+
+    [JsonIgnore]
+    private IEnumerable<IUnlockCondition>? allOrganelleUnlockConditions;
+
+    /// <summary>
+    ///   If true, <see cref="IsUnlocked"/> will always return true
+    /// </summary>
+    [JsonProperty]
+    public bool UnlockAll { get; set; }
+
+    [JsonProperty]
+    public GameWorld World { get; set; } = null!;
+
+    public static Type[] GetSupportedStageTypes()
+    {
+        return new[]
+        {
+            typeof(MicrobeStage),
+        };
+    }
+
+    public void InitUnlockConditionTracking()
+    {
+        GetAllUnlockConditions();
+
+        foreach (var constraint in allOrganelleUnlockConditions!)
+            constraint.OnInit(World.StatisticsTracker);
```

Comment:
Am I right in assuming that it seems like the unlock condition objects from the registry are directly used here and their state changes? If I'm correct then this is a pretty bad architecture problem. Any registry loaded type should be immutable. Instead the design should be "inverted" so that whenever the constraints are checked the statistics tracker needs t obe passed in as a parameter there.